Query raster using supplied GeoJSON

mmcfarland opened the pull request:
#### Overview

Builds off of the naive implementation of `/counts` in #2. 

Add required GeoJSON polygon payload to `/counts` endpoint.  The provided geometry is used to generate a windowed read on the specified raster.  Only the raster cell values which intersect with the geometry are used in the count analysis.

Switches from a slow iteration of array elements to a fast numpy implementation.
#### Testing

You'll need a raster in EPSG:5070 projection to test this PR, instructions are detailed in updates to the README in 4972a72.  GeoJSON can be generated at http://geojson.io, but very large areas may introduce a `MemoryError`

Commit:
Allow count query to use input polygon

Count endpoint accepts a valid GeoJSON object defining an area of
interest to compute count statistics for.  Also improves the method of
making counts, improving performance considerably.

mmcfarland committed:
Add scipy dependency

Not ultimately used at this time, but has many numpy masked array stats
functions.  Nice to have available to play around with capabilities,
even though it's not currently required for the code to execute.

mmcfarland committed:
Update README

New instructions for requesting count, plus link to representative
sample data.

mmcfarland committed:
Move request info to JSON config

Consolidates the request information into a single json payload.  This
anticipates being able to specify multiple rasters to operate on and
additional parameters.  This format is similar to that used in other
team projects.

hectcastro commented on src/geop/main.py:
```diff
@@ -28,26 +33,48 @@ def count():
+        all_touched=True
+    )
+
+    masked_data = np.ma.array(data=data, mask=geom_mask.astype(bool))
```
Can this procedure be achieved with `rasterio.mask.mask`? A little out of my depth here, but it seems like that function can help mask the area outside of the input shapes with `nodata`. Seems similar to what we're trying to do here.

mmcfarland replied:
Yeah, it looks like it can - had not come across that method.  I'll give it a try and compare to how it performs against this implementation.

hectcastro commented on src/geop/main.py:
```diff
@@ -28,26 +33,48 @@ def count():
+    values, counts = np.unique(masked_data.compressed(), return_counts=True)
+
+    # Make dict of val: count with string keys for valid json
+    count_map = dict(zip(values, counts))
```
Since it doesn't really matter what type `values` is at this stage, it might make sense to convert it before the `zip` so that the `format_counts` step below isn't needed:

``` python
count_map = dict(zip(map(str, values), counts))
```

mmcfarland replied:
Good call, I'll combine the two.

hectcastro commented on src/geop/utils.py:
```diff
@@ -0,0 +1,64 @@
+    calculates the affine transformation needed to map the coordinates of the
+    geometry onto a resulting array defined by the shape of the window.
+
+    Args:
```
The `raster_src` argument is not documented here.

hectcastro commented on src/geop/utils.py:
```diff
@@ -0,0 +1,64 @@
+        Shapely Geometry with coordinates transformed to the desired srs
+
+    """
+    projection = partial(
```
Nice.

hectcastro commented on src/requirements.txt:
```diff
@@ -1,4 +1,6 @@
 numpy==1.11.1
 rasterio==0.36.0
 requests==2.10.0
+scipy==0.18
```
Is this being used anywhere?

mmcfarland replied:
It's not, but I left a note about it in the commit message which added it: 6e5a3644340cfc8e82118031aecc2e13545f3a5a.  Basically, I keep trying scipy functions, but don't end up implementing with them.  It's nice to have the library included to experiment without needing to rebuild the container.

mmcfarland committed:
Add rasterio mask method

An attempt to use the built in `rasterio.mask.mask` method to mask out
portional of a raster from the supplied geojson polygon.  This
implementation appears to increase execution time by about 50% over the
existing implementation.  I believe this is because the method actually
fills the raster with the `nodata` value, which then needs to be
converted to a numpy masked array in order for fast array
manipulation.

This commit if mostly for posterity and will remove in the future.

mmcfarland commented:
I made the minor clean up improvements, and added an experimental commit to switch to using `rasterio.mask.mask` in     2543b04.  

My basic finding is that the built in function does not return a numpy masked array, meaning that nodata values (as [filled by mask](https://github.com/mapbox/rasterio/blob/96e63f02b5929b43db56a79e06efd748b8d89fa9/rasterio/mask.py#L89-L92)) are written to the raster object.  This extra step seems to increase the execution time by about 50% in my tests.  The actual count results were identical though, so it does supply some confirmation that the other method is also correct. 

I think the technique is interesting and worth knowing about because it may prove valuable when wanting to continue working on an `ndarray` within rasterio and not directly going to the numpy analysis.  I've added and reverted the commit in this PR because I'd like it in the history for reference.

 Here are representative runs with both `?rio_mask` set to False (a) and True (b).

##### a:

``` json
{
  "cellCount": 49121537,
  "counts": {
    "11": 444765,
    "21": 3602498,
    "22": 1543266,
    "23": 641996,
    "24": 253904,
    "31": 299498,
    "41": 28687098,
    "42": 1524283,
    "43": 1164965,
    "52": 124729,
    "71": 456088,
    "81": 6736832,
    "82": 3555144,
    "90": 69665,
    "95": 16806
  },
  "time": 2.2392600000000003
}
```

##### b:

``` json
{
  "cellCount": 49121537,
  "counts": {
    "11": 444765,
    "21": 3602498,
    "22": 1543266,
    "23": 641996,
    "24": 253904,
    "31": 299498,
    "41": 28687098,
    "42": 1524283,
    "43": 1164965,
    "52": 124729,
    "71": 456088,
    "81": 6736832,
    "82": 3555144,
    "90": 69665,
    "95": 16806
  },
  "time": 3.392034
}
```

mmcfarland committed:
Revert "Add rasterio mask method"

This reverts commit 2543b04c2b30f26678fade0b4de3e06285f28f64.

The performance penalty as mentioned in the original commit message is
preventing me from pursing this implementation.

hectcastro commented:
Disappointing, but also interesting that `rasterio.mask.mask()` calls `rasterio.features.geometry_mask()`, which appears to do the same as we do with `rasterize()` (also doesn't appear to write in the fill and returns a np array of type `bool`). Maybe we should use `rasterio.features.geometry_mask()` then?

mmcfarland commented:
Yeah, I switched to the `geometry_mask` function in https://github.com/azavea/simple-raster-processing/pull/12/commits/fb0bca758d08a69af850198ff4a5015be2e4e5e6 (a different PR).  However, I suspect there may be more to `mask.mask()` than I've been able to discover so far.  I'm going to keep investigating some, especially with an eye towards understanding how rasterio navigates numpy masks vs GDAL masks (https://mapbox.github.io/rasterio/masks.html)

Commit:
Merge pull request #7 from azavea/feature/mjm/config-req

Move request info to JSON config
